File: src/exoticatechnologies/modifications/upgrades/methods/RecoverMethod.java
package exoticatechnologies.modifications.upgrades.methods;

import com.fs.starfarer.api.Global;
import com.fs.starfarer.api.campaign.*;
import com.fs.starfarer.api.campaign.econ.MarketAPI;
import com.fs.starfarer.api.combat.ShipAPI;
import com.fs.starfarer.api.fleet.FleetMemberAPI;
import com.fs.starfarer.api.impl.campaign.ids.Commodities;
import com.fs.starfarer.api.util.Misc;
import exoticatechnologies.campaign.rulecmd.ETInteractionDialogPlugin;
import exoticatechnologies.hullmods.ExoticaTechHM;
import exoticatechnologies.modifications.ShipModifications;
import exoticatechnologies.modifications.upgrades.Upgrade;
import exoticatechnologies.util.StringUtils;
import exoticatechnologies.util.Utilities;

import java.util.Map;

public class RecoverMethod implements UpgradeMethod {
    @Override
    public String getOptionText(FleetMemberAPI fm, ShipModifications mods, Upgrade upgrade, MarketAPI market) {
        int creditCost = getCreditCost(fm, mods, upgrade);

        String creditCostFormatted = Misc.getFormat().format(creditCost);
        return StringUtils.getTranslation("UpgradeMethods", "RecoverOption")
                .format("credits", creditCostFormatted)
                .toString();
    }

    @Override
    public String getOptionTooltip(FleetMemberAPI fm, ShipModifications es, Upgrade upgrade, MarketAPI market) {
        return StringUtils.getTranslation("UpgradeMethods", "RecoverOptionTooltip").toString();
    }

    @Override
    public boolean canShow(FleetMemberAPI fm, ShipModifications mods, Upgrade upgrade, MarketAPI market) {
        return mods.hasUpgrade(upgrade);
    }

    @Override
    public boolean canUse(FleetMemberAPI fm, ShipModifications mods, Upgrade upgrade, MarketAPI market) {
        if (mods.hasUpgrade(upgrade)) {
            int creditCost = getCreditCost(fm, mods, upgrade);
            return Global.getSector().getPlayerFleet().getCargo().getCredits().get() >= creditCost;
        }
        return false;
    }

    @Override
    public void apply(InteractionDialogAPI dialog, ETInteractionDialogPlugin plugin, ShipModifications mods, Upgrade upgrade, MarketAPI market, FleetMemberAPI fm) {
        TextPanelAPI textPanel = dialog.getTextPanel();
        CampaignFleetAPI fleet = Global.getSector().getPlayerFleet();

        ShipAPI.HullSize hullSize = fm.getHullSpec().getHullSize();

        int creditCost = getCreditCost(fm, mods, upgrade);

        CargoStackAPI stack = Utilities.getUpgradeChip(fleet.getCargo(), upgrade.getKey(), mods.getUpgrade(upgrade));
        if (stack != null) {
            stack.add(1);
        } else {
            fleet.getCargo().addSpecial(upgrade.getNewSpecialItemData(mods.getUpgrade(upgrade)), 1);
        }

        mods.removeUpgrade(upgrade);
        fleet.getCargo().getCredits().subtract(creditCost);

        mods.save(fm);
        ExoticaTechHM.addToFleetMember(fm);

        textPanel.addParagraph(StringUtils.getString("UpgradesDialog", "UpgradeRecoveredSuccessfully"));
    }

    /**
     * Sums up the floats in the map.
     *
     * @param resourceCosts
     * @return The sum.
     */
    private static int getCreditCostForResources(Map<String, Integer> resourceCosts) {
        float creditCost = 0;

        for (Map.Entry<String, Integer> resourceCost : resourceCosts.entrySet()) {
            creditCost += Utilities.getItemPrice(resourceCost.getKey()) * resourceCost.getValue();
        }
        return (int) creditCost;
    }

    private static int getCreditCost(FleetMemberAPI fm, ShipModifications mods, Upgrade upgrade) {
        float resourceCreditCost = getCreditCostForResources(upgrade.getResourceCosts(fm, mods.getUpgrade(upgrade)));
        int creditCost = (int) (resourceCreditCost * 0.166);

        return creditCost;
    }

    @Override
    public void modifyResourcesPanel(ETInteractionDialogPlugin plugin, Map<String, Float> resourceCosts, FleetMemberAPI fm, Upgrade upgrade, boolean hovered) {
        if (hovered) {
            resourceCosts.put(Commodities.CREDITS, (float) getCreditCost(fm, plugin.getExtraSystems(), upgrade));
        }
    }
}

File: src/exoticatechnologies/ui/impl/shop/exotics/methods/RecoverMethod.kt
package exoticatechnologies.ui.impl.shop.exotics.methods

import com.fs.starfarer.api.Global
import com.fs.starfarer.api.campaign.CampaignFleetAPI
import com.fs.starfarer.api.campaign.econ.MarketAPI
import com.fs.starfarer.api.combat.ShipVariantAPI
import com.fs.starfarer.api.fleet.FleetMemberAPI
import exoticatechnologies.modifications.ShipModifications
import exoticatechnologies.modifications.bandwidth.Bandwidth
import exoticatechnologies.modifications.exotics.Exotic
import exoticatechnologies.modifications.exotics.types.ExoticType
import exoticatechnologies.util.StringUtils
import exoticatechnologies.util.Utilities

class RecoverMethod : DestroyMethod() {
    override val key: String = "recover"

    override fun apply(
        member: FleetMemberAPI,
        variant: ShipVariantAPI,
        mods: ShipModifications,
        exotic: Exotic,
        market: MarketAPI?
    ): String {
        val fleet: CampaignFleetAPI = Global.getSector().playerFleet
        
        val exoticData = mods.getExoticData(exotic)
        val exoticType = exoticData?.type ?: ExoticType.NORMAL

        val stack = Utilities.getExoticChip(fleet.cargo, exotic.key, exoticType.nameKey)
        if (stack != null) {
            stack.add(1f)
        } else {
            fleet.cargo.addSpecial(exotic.getNewSpecialItemData(exoticType), 1f)
        }

        Global.getSector().playerStats.storyPoints--

        /**
         * Super call
         */
        super.apply(member, variant, mods, exotic, market)

        return StringUtils.getString("ExoticsDialog", "ExoticRecovered")
    }

    override fun canUse(member: FleetMemberAPI, mods: ShipModifications, exotic: Exotic, market: MarketAPI?): Boolean {
        if (Global.getSector().playerStats.storyPoints >= 1f) {
            return super.canUse(member, mods, exotic, market)
        }
        return false
    }

    override fun getButtonText(exotic: Exotic): String {
        return StringUtils.getString("ExoticsDialog", "RecoverExotic")
    }

    override fun getResourceMap(
        member: FleetMemberAPI,
        mods: ShipModifications,
        exotic: Exotic,
        market: MarketAPI?,
        hovered: Boolean
    ): Map<String, Float>? {
        if (hovered) {
            val resourceCosts: MutableMap<String, Float> = mutableMapOf()
            resourceCosts[Utilities.STORY_POINTS] = 1f

            val stack = Utilities.getExoticChip(Global.getSector().playerFleet.cargo, exotic.key)
            if (stack != null) {
                resourceCosts.put(Utilities.formatSpecialItem(stack.specialDataIfSpecial), -1f)
            } else {
                resourceCosts.put(
                    "&" + StringUtils.getTranslation("ShipListDialog", "ChipName")
                        .format("name", exotic.name)
                        .toStringNoFormats(), -1f)
            }

            if (exotic.getExtraBandwidth(member, mods, mods.getExoticData(exotic)) > 0) {
                resourceCosts[Bandwidth.BANDWIDTH_RESOURCE] = -exotic.getExtraBandwidth(
                    member,
                    mods,
                    mods.getExoticData(exotic)
                )
            }

            return resourceCosts
        }
        return null
    }
}